core/services/ocrcommon/adapters.go

```diff
@@ -81,47 +87,120 @@ func (c *OCR3ContractTransmitterAdapter) FromAccount() (ocrtypes.Account, error)
 
 var _ ocr3types.OnchainKeyring[[]byte] = (*OCR3OnchainKeyringMultiChainAdapter)(nil)
 
+func MarshalMultichainKeyBundle(ost map[string]ocr2key.KeyBundle) (ocrtypes.OnchainPublicKey, error) {
+	pubKeys := map[string]ocrtypes.OnchainPublicKey{}
+	for k, b := range ost {
+		pubKeys[k] = []byte(b.PublicKey())
+
+	}
+	return MarshalMultichainPublicKey(pubKeys)
+}
+
+func MarshalMultichainPublicKey(ost map[string]ocrtypes.OnchainPublicKey) (ocrtypes.OnchainPublicKey, error) {
+	var pubKeys [][]byte
+	for k, pubKey := range ost {
+		typ, err := chaintype.ChainType(k).Type()
+		if err != nil {
+			// skipping unknown key type
+			continue
+		}
+		buf := new(bytes.Buffer)
+		binary.Write(buf, binary.LittleEndian, typ)
+		binary.Write(buf, binary.LittleEndian, uint16(len(pubKey)))
+		_, _ = buf.Write(pubKey)
+		pubKeys = append(pubKeys, buf.Bytes())
+	}
+	// sort keys based on encoded type to make encoding deterministic
+	slices.SortFunc(pubKeys, func(a, b []byte) int { return cmp.Compare(a[0], b[0]) })
+	return bytes.Join(pubKeys, nil), nil
+}
+
+func UnmarshalMultichainPublicKey(d []byte) (map[string]ocrtypes.OnchainPublicKey, error) {
+	m := map[string]ocrtypes.OnchainPublicKey{}
+	buf := bytes.NewReader(d)
+
+	for {
+		// type
+		typ, err := buf.ReadByte()
+		if err != nil {
+			return nil, err
+		}
+		// length
+		var len uint16
+		err = binary.Read(buf, binary.LittleEndian, &len)
+		if err != nil {
+			return nil, err
+		}
+		// value
+		pubKey := make([]byte, len)
+		n, err := buf.Read(pubKey)
+		if err != nil {
+			return nil, err
+		}
+		if n != int(len) {
+			return nil, io.EOF
+		}
+
+		k, err := chaintype.NewChainType(typ)
+		if err != nil {
+			// skipping unknown key type
```

Comment:
I'm deliberately skipping unknown keys, that way if the contract is configured with chain types not yet known to the node it'll keep working